From db2ab377f10db5d3cca2bdec40015d4646a46882 Mon Sep 17 00:00:00 2001
Message-Id: <db2ab377f10db5d3cca2bdec40015d4646a46882.1728389216.git.mykyta_poturai@epam.com>
In-Reply-To: <bc35a1559ba18018db00e017ef9359f091274151.1728389216.git.mykyta_poturai@epam.com>
References: <bc35a1559ba18018db00e017ef9359f091274151.1728389216.git.mykyta_poturai@epam.com>
From: Oleksandr Tyshchenko <oleksandr_tyshchenko@epam.com>
Date: Mon, 15 Jul 2019 12:06:20 +0300
Subject: [PATCH 16/22] xen_back: Report that Guest is online before switching
 state to connected

To avoid possible races when Guest has just detected that Host had gone
into connected state and unblocks pvz_completion to proceed further
(map phys heap), but Host hasn't detected Guest's state change yet
and Guest is still treated as offline at the Host side.

Signed-off-by: Oleksandr Tyshchenko <oleksandr_tyshchenko@epam.com>
---
 services/system/rogue/common/env/xen/xen_back.c | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/services/system/rogue/common/env/xen/xen_back.c b/services/system/rogue/common/env/xen/xen_back.c
index 7613273..f3f61c9 100644
--- a/services/system/rogue/common/env/xen/xen_back.c
+++ b/services/system/rogue/common/env/xen/xen_back.c
@@ -603,6 +603,8 @@ static void xdrv_fe_on_changed(struct xenbus_device *xb_dev,
 	case XenbusStateReconfigured:
 		/* fall through */
 	case XenbusStateInitWait:
+		/* fall through */
+	case XenbusStateConnected:
 		break;
 
 	case XenbusStateUnknown:
@@ -632,12 +634,9 @@ static void xdrv_fe_on_changed(struct xenbus_device *xb_dev,
 			break;
 		}
 
-		xenbus_switch_state(xb_dev, XenbusStateConnected);
-		break;
-
-	case XenbusStateConnected:
 		xen_back_pvz_connection->sVmmFuncTab.pfnOnVmOnline(
 			drv_info->osid);
+		xenbus_switch_state(xb_dev, XenbusStateConnected);
 		break;
 	}
 }
-- 
2.34.1

